### ISSUE TITLE
texture reconstruction

### ISSUE THREAD

**Issue by tejank10** [ShichenLiu/SoftRas#44]
Hello @ShichenLiu @chenweikai ,

Could you please provide some details about texture reconstruction? I am primarily interested in the following points:
- `texture_type` used for generating texture
- Color palette size used
- Definition of the Laplacian loss for color

Thanks.

**Comment by ShichenLiu** [comment]
Hi there,

In our experiments we used surface texture type and the Laplacian neighbor is set to the adjacent triangles in order to compute the Laplacian loss on color. We used about 10~20 as color palette size. We think relatively small size could impose the networks to learn better concept of color palette.

**Comment by tejank10** [comment]
Thank you very much @ShichenLiu for clarifying the details. The texture generated is directly passed on to the renderer, or does it undergo transformation (like in case of generated displacement, which is then added to vertices)?

**Comment by ShichenLiu** [comment]
The texture generated is directly passed to the renderer since it is
"picked" from the source images.

- Shichen

**Comment by zyz-1998** [comment]
> Thank you very much @ShichenLiu for clarifying the details. The texture generated is directly passed on to the renderer, or does it undergo transformation (like in case of generated displacement, which is then added to vertices)?

Hi, did you implement the texture reconstruction module of the project, would you mind sharing it for learning, looking forward to your reply

**Comment by tejank10** [comment]
I have created a gist [here](https://gist.github.com/tejank10/a116654e644be371fa7296dfc185f73a) regarding it. The module and function are to be placed in models.py

**Comment by zyz-1998** [comment]
Thank you very much for your reply and help

**Comment by zyz-1998** [comment]
> I have created a gist [here](https://gist.github.com/tejank10/a116654e644be371fa7296dfc185f73a) regarding it. The module and function are to be placed in models.py

hi, sorry to disturb you, you can share the color loss code, looking forward to your reply,

**Comment by tejank10** [comment]
I have updated the gist with Texture Laplacian Loss, I assume that's what you mean by the color loss code.
Other than this, there is L1 loss over rendered vs input RGB image whichis also a color loss.

**Comment by zyz-1998** [comment]
I get it, thank you very much

**Comment by zyz-1998** [comment]
> I have updated the gist with Texture Laplacian Loss, I assume that's what you mean by the color loss code.
> Other than this, there is L1 loss over rendered vs input RGB image whichis also a color loss.

Thanks again for your help, I have initially achieved texture reconstruction, but there are some problems with the effect, the textures I get are almost all black and white tones, here are my [train.py](https://github.com/zyz-1998/project/blob/master/train.py) and [model.py](https://github.com/zyz-1998/project/blob/master/models.py) , can help me see my settings there Is there something wrong, looking forward to your reply!

**Comment by tejank10** [comment]
Can you put a couple of pictures of the textures you were getting?

**Comment by zyz-1998** [comment]
> Maybe [this](https://github.com/zyz-1998/project/blob/f539f9be4511085b54665292828a94cc2ab210dd/train.py#L127) is the culprit, the losses is taken wrt `images_a[-1]` which is a single image. I have updated the gist with L1 loss that I had used. It is analogous to the `multiview_iou_loss` which is in [losses.py](https://github.com/ShichenLiu/SoftRas/blob/master/soft_renderer/losses.py) of SoftRas.

Thank you very much. At present, the texture reconstruction effect is good. I believe the code you contributed on texture reconstruction will help more people.

**Comment by xiehousen** [comment]
> > Maybe [this](https://github.com/zyz-1998/project/blob/f539f9be4511085b54665292828a94cc2ab210dd/train.py#L127) is the culprit, the losses is taken wrt `images_a[-1]` which is a single image. I have updated the gist with L1 loss that I had used. It is analogous to the `multiview_iou_loss` which is in [losses.py](https://github.com/ShichenLiu/SoftRas/blob/master/soft_renderer/losses.py) of SoftRas.
> 
> Thank you very much. At present, the texture reconstruction effect is good. I believe the code you contributed on texture reconstruction will help more people.

hello,  sorry to disturb you,  can you share the code of model.py? 
I meet the same question as you when  I recompile the loss and put the col_gen in model.py. Look forward to your reply

**Comment by zyz-1998** [comment]
> ![demo_0003000](https://user-images.githubusercontent.com/37787619/98779229-8564bf80-242e-11eb-9116-fe7716ef12d7.png)
> 
> > > Maybe [this](https://github.com/zyz-1998/project/blob/f539f9be4511085b54665292828a94cc2ab210dd/train.py#L127) is the culprit, the losses is taken wrt `images_a[-1]` which is a single image. I have updated the gist with L1 loss that I had used. It is analogous to the `multiview_iou_loss` which is in [losses.py](https://github.com/ShichenLiu/SoftRas/blob/master/soft_renderer/losses.py) of SoftRas.
> > 
> > 
> > Thank you very much. At present, the texture reconstruction effect is good. I believe the code you contributed on texture reconstruction will help more people.
> 
> hello, sorry to disturb you, can you share the code of model.py?
> I meet the same question as you when I recompile the loss and put the col_gen in model.py. Look forward to your reply

color_loss = (l1loss(predicts[0][0:,:3],targets_a[:,:3])+\
l1loss(predicts[1][0:,:3],targets_a[:,:3])+\
l1loss(predicts[2][0:,:3],targets_b[:,:3])+\
l1loss(predicts[3][0:,:3],targets_b[:,:3])) / 4

**Comment by ShichenLiu** [comment]
Provided samples of reconstructed color meshes in the repo!

**Comment by zoezhou1999** [comment]
Hello, when I used the code provided in this issue to reconstruct texture and shape, the texture remains black. is it normal?
or the code is designed to optimize texture only by providing already-good-looking mesh? Thank you!

**Comment by zoezhou1999** [comment]
Hi @ShichenLiu 
I followed the code snippets provided by @tejank10 . The texture reconstruction seems to work. But I noticed that the current texture reconstruction code and the descriptions in paper supplementary materials generate batch_size x face_num x 1 x 3 textures but the saving obj of SoftRas requires texture_res>=2, so I just repeat the 2nd dim. But I think it may affect the performance? So do you mind providing some hints on your texture reconstruction implementation about this? Thank you!
